src/zelos/hooks.py

```diff
@@ -287,6 +294,69 @@ def syscall_callback_wrapper(zelos, sysname, args, retval):
             )
         return self._add_zelos_hook(syscall_hook_type, callback, name)
 
+    def attempt_to_setup_func_hooks(self):
+        """
+        Attempt to get the function base. If unsuccessful, place a
+        hook that waits for the module to be mapped.
+        """
+        base_address = self.z.memory.get_module_base(self.z.target_binary_path)
+        if base_address is not None:
+            self.logger.verbose("Setting up func hooks")
+            self._func_hooks_enabled = True
+            for (name, callback) in self._func_hooks_to_register:
+                self.register_func_hook(name, callback)
+            self._func_hooks_to_register = None
+            return
+        self.logger.verbose("Delaying func hook setup")
+
+        # Wait for the main module to be loaded before setting up the
+        # function hooks
+        def main_module_is_loaded():
+            base_address = self.z.memory.get_module_base(
+                self.z.target_binary_path
+            )
+            return base_address is not None
+
+        def delayed_func_hook_setup(zelos, access, address, size, data):
+            if main_module_is_loaded():
+                self.attempt_to_setup_func_hooks()
+
+        self.register_mem_hook(
+            HookType.MEMORY.INTERNAL_MAP,
+            delayed_func_hook_setup,
+            end_condition=main_module_is_loaded,
+        )
+
+    def register_func_hook(self, name: str, callback) -> HookInfo:
+        address_ptr = self.z.main_module._elf_dynamic_import_addrs.get(
+            name, None
+        )
+        if address_ptr is None:
+            self.logger.notice(f"Couldn't find function {name}")
+            return None
+
+        if not self._func_hooks_enabled:
+            self._func_hooks_to_register.append((name, callback))
+            return
+
+        address_ptr += self.z.memory.get_module_base(self.z.target_binary_path)
+
+        def read_wrapper(z, access, address, size, value):
+            callback(z)
+
+        # TODO: Another way to do this is to wait for these pointers
+        # to be written to, and then hook those execution locations.
```

Comment:
Would this work by creating a __write hook__ on `address_ptr`, the callback of which would create the __exec hook__   (just like the inline comment below)?

Reply:
Resolved offline.